### ISSUE TITLE
SCRAM-SHA-1(-PLUS) + SCRAM-SHA-256(-PLUS) supports

### ISSUE THREAD

**Issue by Neustradamus** [svarshavchik/courier#19]
"When using the SASL SCRAM mechanism, the SCRAM-SHA-256-PLUS variant SHOULD be preferred over the SCRAM-SHA-256 variant, and SHA-256 variants [RFC7677] SHOULD be preferred over SHA-1 variants [RFC5802]".

Can you add support for?
- SCRAM-SHA-1(-PLUS):
-- https://tools.ietf.org/html/rfc5802
-- https://tools.ietf.org/html/rfc6120

- SCRAM-SHA-256(-PLUS):
-- https://tools.ietf.org/html/rfc7677 since 2015-11-02
-- https://tools.ietf.org/html/rfc8600 since 2019-06-21: https://mailarchive.ietf.org/arch/msg/ietf-announce/suJMmeMhuAOmGn_PJYgX5Vm8lNA

I add SCRAM-SHA-512(-PLUS): https://xmpp.org/extensions/inbox/hash-recommendations.html

Linked to:
- https://github.com/scram-xmpp/info/issues/1

**Comment by svarshavchik** [comment]
I'm open to reviewing patches and pull requests, of course.

**Comment by Neustradamus** [comment]
@svarshavchik: Have you looked?
Cyrus SASL has it, ...
Dovecot too...

**Comment by svarshavchik** [comment]
No, I do not have an update here. In the last year noone else mentioned any need for this, and if someone needs it bad enough, patches are still welcome.

**Comment by Neustradamus** [comment]
@svarshavchik: I wish you a Happy New Year!

Have you progressed on it?

Thanks in advance.

**Comment by svarshavchik** [comment]
CRAM and hash-based authentication schemes were preferred a long time ago before the overall TLS encryption became widespread and available. Today TLS offers a far superior level of security than password-only schemes.

Additionally most of these require the server to store client passwords in plain text, which is actually even worse from security perspective. The most secure password validation mechanism is using TLS for the entire client session, with a plain text password that's one-way hashed on the server, so even if the server is compromised the passwords will not be.

If someone were to submit a patch to implement these I'll review it, but this is not a priority for me.

**Comment by Neustradamus** [comment]
@svarshavchik: Thanks for your reply but Courier must not to support obsolete and unsecure CRAM-MD5 and DIGEST-MD5:
- https://datatracker.ietf.org/doc/html/rfc6331
- https://datatracker.ietf.org/doc/html/draft-zeilenga-luis140219-crammd5-to-historic-00

It is really very important, to add, for example in first time without TLS Binding (= without -PLUS variants).

It is already done in:
- Cyrus SASL / Cyrus IMAP
- Dovecot
- Exim
- Postfix with Cyrus SASL or Postfix with Dovecot SASL
- msmtp / mpop
- DataEnter CryptoFilter / DataEnter POPBeamer / DataEnter SMTPBeamer / DataEnter XWall
- NeoMutt
- SquirelMail
- SnappyMail
- MimeKit / MailKit
- ...

A more complete list here: https://github.com/scram-xmpp/info/issues/1.

Thanks in advance.

Linked to:
- https://github.com/svarshavchik/courier/issues/20

**Comment by svarshavchik** [comment]
CRAM-MD5 and other CRAM authentication methods are disabled by default, and require explicit configuration in order to enable.

As such, no action is needed, since CRAM-MD5 is not enabled in a default configuration.

**Comment by Neustradamus** [comment]
@svarshavchik: Thanks for your clarification but badly some admins enable it...

The SCRAM supports are important, it is for this I relaunch you about it.

**Comment by svarshavchik** [comment]
In the interim I can update the documentation to note that MD5 should only be enabled for legacy client support.

Like I wrote earlier, if someone needs this sufficiently to implement a patch I'll review it, but this is not a very high priority since HMAC/SCRAM authentication mechanism are really no longer needed, and TLS is better than HMAC/SCRAM. Additionally, Courier supports certificate-based authentication, without any passwords, which is the highest level of secuirty, overall. Implementing SCRAM is mostly a wasted effort. High-security installations will use certificates to authenticate, anyway.

**Comment by Neustradamus** [comment]
@svarshavchik: Yes you can add about MD5.

Maybe it is a good page about SCRAM: https://techcommunity.microsoft.com/t5/azure-database-for-postgresql/how-to-securely-authenticate-with-scram-in-postgres-13/ba-p/1548319.

**Comment by Neustradamus** [comment]
@svarshavchik: I have added the last RFC in the description: RFC9051: Internet Message Access Protocol (IMAP) - Version 4rev2:
- https://tools.ietf.org/html/rfc9051

I wish you a good reading.

**Comment by svarshavchik** [comment]
imap4rev2 lists the SCRAM authentication methods as SHOULD, not MUST, they're not required. In any case, Courier does not implement imap4rev2, anyway.
